safenode/src/node/mod.rs

```diff
@@ -0,0 +1,132 @@
+mod api;
+mod error;
+mod event;
+
+pub use event::NodeEvent;
+use libp2p::request_response::ResponseChannel;
+
+use self::{
+    error::{Error, Result},
+    event::NodeEventsChannel,
+};
+use crate::{
+    network::{
+        Cmd, CmdResponse, Network, NetworkEvent, NetworkSwarmLoop, Query, QueryResponse, Request,
+        Response,
+    },
+    storage::{chunks::ChunkAddress, DataStorage},
+};
+use assert_fs::TempDir;
+use futures::StreamExt;
+use std::time::Duration;
+use tokio::task::spawn;
+
+/// Safe node
+#[derive(Clone)]
+pub struct Node {
+    network: Network,
+    storage: DataStorage,
+    node_events_channel: NodeEventsChannel,
+}
+
+impl Node {
+    /// Create and run the `Node`
+    pub async fn run() -> Result<(Self, NodeEventsChannel)> {
+        let (network, mut network_events, network_event_loop) = NetworkSwarmLoop::new()?;
+        let temp_dir =
+            TempDir::new().map_err(|_| Error::Other("failed to create temp_dir".to_string()))?;
+        let storage = DataStorage::new(&temp_dir);
+        let node_events_channel = NodeEventsChannel::default();
+        let node = Self {
+            network,
+            storage,
+            node_events_channel: node_events_channel.clone(),
+        };
+        let mut node_clone = node.clone();
+
+        // Spawn the network task for it to run in the background.
+        let _ = spawn(network_event_loop.run());
+
+        let _ = spawn(async move {
+            loop {
+                let event = match network_events.next().await {
+                    Some(event) => event,
+                    None => {
+                        tokio::time::sleep(Duration::from_millis(10)).await;
+                        continue;
+                    }
+                };
+                if let Err(err) = node_clone.handle_network_events(event).await {
+                    warn!("Error while handling network events: {err}");
+                }
+            }
+        });
+
+        Ok((node, node_events_channel))
+    }
+
+    /// Handle incoming `NetworkEvent`
+    async fn handle_network_events(&mut self, event: NetworkEvent) -> Result<()> {
+        match event {
+            NetworkEvent::RequestReceived { req, channel } => {
+                self.handle_request(req, channel).await?
+            }
+            NetworkEvent::PeerAdded => {
+                self.node_events_channel
+                    .broadcast(NodeEvent::ConnectedToNetwork);
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Handle incoming `Request` from a peer
+    async fn handle_request(
+        &mut self,
+        request: Request,
+        response_channel: ResponseChannel<Response>,
+    ) -> Result<()> {
+        trace!("Handling request: {request:?}");
+        match request {
+            Request::Query(query) => match query {
+                Query::GetChunk(xor_name) => {
+                    let addr = ChunkAddress(xor_name);
+                    let chunk = self.storage.query(&addr).await?;
```

Comment:
the self storage shall be just a read file from local disk.
so a `get` call instead of `query` shall be more precise ?

Also, I am wondering for the self storage, a direct read/write file shall be enough, no need to be `async` ?

Reply:
also, the result of the data fetch may be an error.
in that case, the error shall be sent back as a response.
shall not just `unwrap` directly, which will result in a `terminated flow` and cause the request sender have to `time out`